feat: add php language flavors and increase test coverage

urbit-pilled opened the pull request:
/claim #48 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced PHP support with the introduction of `PhpOnly` language variant for specific PHP code handling.
  - Updated PHP grammar rules to improve parsing accuracy and integration with HTML.

- **Bug Fixes**
  - Adjusted PHP grammar and metavariable definitions to fix issues in code parsing and pattern matching.

- **Refactor**
  - Renamed PHP-related language variants for better clarity in the system.

- **Tests**
  - Added new test scenarios for PHP code patterns to ensure transformation robustness.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

morgante commented on resources/metavariable-grammars/php-common-metavariable-grammar.js:
```diff
@@ -318,7 +323,7 @@ module.exports = function defineGrammar(dialect) {
       enum_case: $ => seq(
         optional(field('attributes', $.attribute_list)),
         keyword('case'),
-        field('name', choice($.name, $.grit_metavariable)),
```
All these removals seem wrong - what's the reason?

urbit-pilled replied:
I changed the `$.name` rule to `choice($._name, $.grit_metavariable)` so the choice is no longer necessary everywhere else. 
https://github.com/getgrit/gritql/pull/146/commits/aa4017ca89fe592d9a6c5df19e747bc53fcd7618#diff-d8bdb1fe5a96b8d2e4b8ad3756d3a053d4103b57ae16ad83f8ed2f7a84d33b90L1562-R1566

morgante commented on crates/language/src/target_language.rs:
```diff
@@ -227,7 +232,11 @@ impl PatternLanguage {
+            "php" => match flavor {
+                Some("html") => Some(Self::Php),
+                Some("only") => Some(Self::PhpOnly),
+                _ => Some(Self::PhpOnly),
```
I think we might want HTML as the default actually, since it seems like the most common thing.

urbit-pilled replied:
ok, I'll change the default to html

morgante commented on crates/language/src/php_only.rs:
```diff
@@ -0,0 +1,118 @@
+    language: &'static TSLanguage,
+}
+// use std::io::Write;
+impl PhpOnly {
```
We should follow the pattern from JS of having some shared x-script utils that get used for both flavors instead of repeating everything.

morgante commented on crates/language/src/php.rs:
```diff
@@ -63,17 +62,21 @@ impl Language for Php {
-            ("class GRIT_CLASS { ", " function GRIT_FN(); }"),
-            (" GRIT_FN(", ") { }"),
-            ("[", "];"),
+            ("<?php ", " ?>"),
```
These should be computable by wrapping every snippet_context_strings from phpOnly with `<?php`

urbit-pilled commented on crates/language/src/xscript_util.rs:
```diff
@@ -147,6 +149,56 @@ pub(crate) fn jslike_check_orphaned(n: NodeWithSource<'_>, orphan_ranges: &mut V
+        ("", "{}"),
+    ];
+    pub static ref PHP_CODE_SNIPPETS: Vec<(&'static str, &'static str)> = {
+        let mut php_tag_modifications:Vec<(&'static str, &'static str)> = PHP_ONLY_CODE_SNIPPETS.
+            clone().
+            into_iter().
+            map(|(s1, s2)| {
+                let owned_str1 = Box::leak(Box::new(format!("<?php {}", s1))) as &'static str;
+                let owned_str2 = Box::leak(Box::new(format!("{} ?>", s2))) as &'static str;
+                (owned_str1, owned_str2)
+            }).collect();
+        php_tag_modifications.extend(vec![("", "")]);
+        php_tag_modifications
+    };
```
doesn't include the extra `("", "")`

urbit-pilled replied:
it is required

urbit-pilled commented on crates/language/src/xscript_util.rs:
```diff
@@ -147,6 +149,56 @@ pub(crate) fn jslike_check_orphaned(n: NodeWithSource<'_>, orphan_ranges: &mut V
+        php_tag_modifications
+    };
+    
+}
```
commited in https://github.com/getgrit/gritql/pull/146/commits/2b6a3b1299ffeed63386d367222640a62a019ef3

urbit-pilled commented:
-added the common php stuff into `xscript_utils` (including `snippet_context_strings`)
-computed `snippet_context_strings` for Php by wrapping the php tags around PhpOnly `snippet_context_strings`
-changed the default from `php(only)` to `php(html)`.
